compiler_opt/rl/generate_bc_trajectories.py

```diff
@@ -809,3 +827,162 @@ def select_best_exploration(
+  size_per_file = total_work // total_write_files
+
+  worker_count = (
+      min(os.cpu_count(), num_workers) if num_workers else os.cpu_count())
+  with worker_manager_class(
+      ModuleWorker,
+      worker_count,
+      obs_action_specs=obs_action_spec,
+  ) as lwm:
+    _, result_futures = buffered_scheduler.schedule_on_worker_pool(
+        action=lambda w, j: w.select_best_exploration(j),
+        jobs=work,
+        worker_pool=lwm,
+    )
+    not_done = result_futures
+    succeeded_idx = 0
+    succeeded: List[Optional[Tuple[Tuple[int, Dict[str, Union[str, float, int]],
+                                         Dict[str, Union[str, float, int]]],
+                                   tf.train.SequenceExample]]] = []
+
+    for written_files in range(total_write_files):
+      written_per_file = 0
+      file_name = f'{0}-{1}-of-{2}.tfrecord'.format(output_file_name,
+                                                    written_files,
+                                                    total_write_files)
+      tf_rec_path = (
+          os.path.join(output_path, file_name)
+          if output_path else contextlib.nullcontext())
+      tfrecord_context = (
+          tf.io.TFRecordWriter(tf_rec_path)
+          if output_path else contextlib.nullcontext())
+
+      with tfrecord_context as tfrecord_writer:
+        time_compiler_start = timeit.default_timer()
+        while not_done or succeeded:
+          (done, not_done) = concurrent.futures.wait(not_done, worker_wait)
+          succeeded.extend(
+              [r for r in done if not r.cancelled() and r.exception() is None])
+          failed = [r for r in done if r.exception() is not None]
+          for f in failed:
+            logging.info('Module failed with: %s', f.exception())
+          total_successful_examples += len(succeeded)
+          total_failed_examples += len(done) - len(succeeded)
+          modules_processed += len(done)
+          while written_per_file < size_per_file:
+            logging.log_every_n_seconds(
+                logging.INFO,
+                ('%d success, %d failed out of %d, modules processed'
+                 ' %d\n timing compiler: %f'),
+                10,
+                total_successful_examples,
+                total_failed_examples,
+                total_work,
+                modules_processed,
+                time_compiler_calls,
+            )
+            if len(succeeded) == 0:
+              break
+            extra, records = succeeded.pop().result()
+            total_profiles_max.append(extra[1])
+            total_profiles_pol.append(extra[2])
+            time_compiler_calls = timeit.default_timer() - time_compiler_start
+            if tfrecord_writer:
+              tfrecord_writer.write(records)
+            written_per_file += 1
+            succeeded_idx += 1
+          if written_per_file >= size_per_file:
+            break
+
+  max_profiles_path = ''
+  pol_profiles_path = ''
+  if profiling_file_path:
```

Comment:
so if `profiling_file_path` isn't given, then what happens with the `open` below?

Reply:
The context is set to contextlib.nullcontext(), so I think nothing happens.

Reply:
would it be cleaner to check here and do the `open` stuff only when `profiling_file_path` is set?

Reply:
Yes, I am addressing this.